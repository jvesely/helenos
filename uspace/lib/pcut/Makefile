#
# This file was generated by call to update-from-master.sh
#

USPACE_PREFIX = ../..

include helenos.mak

include $(USPACE_PREFIX)/Makefile.common

