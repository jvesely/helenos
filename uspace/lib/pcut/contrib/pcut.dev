[Project]
FileName=pcut.dev
Name=PCUT
Type=2
Ver=2
ObjFiles=
Includes=include
Libs=
PrivateResource=
ResourceIncludes=
MakeIncludes=windows.mak;contrib/devcpp.mak
Compiler=-std=c99_@@_
IsCpp=0
OverrideOutputName=pcut.lib
OverrideOutput=1
UseCustomMakefile=0
Folders=Sources
IncludeVersionInfo=0
SupportXPThemes=0
CompilerSet=3
CompilerSettings=0000000100000000000000000
UnitCount=17

[Unit1]
FileName=src\assert.c
CompileCpp=0
Folder=Sources
Compile=1
Link=1
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit2]
FileName=src\helper.c
CompileCpp=0
Folder=Sources
Compile=1
Link=1
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit3]
FileName=src\internal.h
CompileCpp=0
Folder=Sources
Compile=1
Link=1
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit4]
FileName=src\list.c
CompileCpp=0
Folder=Sources
Compile=1
Link=1
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit5]
FileName=src\main.c
CompileCpp=0
Folder=Sources
Compile=1
Link=1
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit6]
FileName=src\print.c
CompileCpp=0
Folder=Sources
Compile=1
Link=1
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit7]
FileName=src\run.c
CompileCpp=0
Folder=Sources
Compile=1
Link=1
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit8]
FileName=src\report\report.c
CompileCpp=0
Folder=Sources
Compile=1
Link=1
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit9]
FileName=src\report\report.h
CompileCpp=0
Folder=Sources
Compile=1
Link=1
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit10]
FileName=src\report\tap.c
CompileCpp=0
Folder=Sources
Compile=1
Link=1
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit11]
FileName=src\report\xml.c
CompileCpp=0
Folder=Sources
Compile=1
Link=1
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit12]
FileName=src\os\generic.c
CompileCpp=0
Folder=Sources
Compile=1
Link=1
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit13]
FileName=src\os\stdc.c
CompileCpp=0
Folder=Sources
Compile=1
Link=1
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit14]
FileName=README.rst
Folder=PCUT
Compile=0
Link=0
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit15]
FileName=windows.mak
Folder=PCUT
Compile=0
Link=0
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit16]
FileName=contrib\devcpp.mak
Folder=PCUT
Compile=0
Link=0
Priority=1000
OverrideBuildCmd=0
BuildCmd=

[Unit17]
FileName=pcut.mak
Folder=PCUT
Compile=0
Link=0
Priority=1000
OverrideBuildCmd=0
BuildCmd=

