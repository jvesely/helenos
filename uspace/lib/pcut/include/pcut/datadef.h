/*
 * Copyright (c) 2012-2013 Vojtech Horky
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * - Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following disclaimer in the
 *   documentation and/or other materials provided with the distribution.
 * - The name of the author may not be used to endorse or promote products
 *   derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/** @file
 * Data types internally used by PCUT.
 */
#ifndef PCUT_DATADEF_H_GUARD
#define PCUT_DATADEF_H_GUARD

#include <pcut/prevs.h>
#include <stdlib.h>
#include <stddef.h>

/** @cond devel */

#if defined(__GNUC__) || defined(__clang__)
#define PCUT_CC_UNUSED_VARIABLE(name, initializer) \
	name __attribute__((unused)) = initializer
#else
#define PCUT_CC_UNUSED_VARIABLE(name, initializer) \
	name = initializer
#endif


enum {
	PCUT_KIND_SKIP,
	PCUT_KIND_NESTED,
	PCUT_KIND_SETUP,
	PCUT_KIND_TEARDOWN,
	PCUT_KIND_TESTSUITE,
	PCUT_KIND_TEST
};

enum {
	PCUT_EXTRA_TIMEOUT,
	PCUT_EXTRA_SKIP,
	PCUT_EXTRA_LAST
};

enum {
	PCUT_MAIN_EXTRA_PREINIT_HOOK,
	PCUT_MAIN_EXTRA_INIT_HOOK,
	PCUT_MAIN_EXTRA_REPORT_XML,
	PCUT_MAIN_EXTRA_LAST
};

/** Generic wrapper for test cases, test suites etc. */
typedef struct pcut_item pcut_item_t;

/** Extra information about a test. */
typedef struct pcut_extra pcut_extra_t;

/** Extra information for the main() function. */
typedef struct pcut_main_extra pcut_main_extra_t;

/** Test method type. */
typedef void (*pcut_test_func_t)(void);

/** Set-up or tear-down method type. */
typedef void (*pcut_setup_func_t)(void);

/** @copydoc pcut_extra_t */
struct pcut_extra {
	/** Discriminator for the union.
	 *
	 * Use PCUT_EXTRA_* to determine which field of the union is used.
	 */
	int type;
	/** Test-specific time-out in seconds. */
	int timeout;
};

/** @copydoc pcut_main_extra_t */
struct pcut_main_extra {
	/** Discriminator for the union.
	 *
	 * Use PCUT_MAIN_EXTRA_* to determine which field of the union is used.
	 */
	int type;
	/** Callback once PCUT initializes itself. */
	void (*init_hook)(void);
	/** Callback even before command-line arguments are processed. */
	void (*preinit_hook)(int *, char ***);
};

/** @copydoc pcut_item_t */
struct pcut_item {
	/** Link to previous item. */
	pcut_item_t *previous;
	/** Link to next item. */
	pcut_item_t *next;

	/** Unique id of this item. */
	int id;

	/** Discriminator for this item. */
	int kind;

	/** Name of this item. */
	const char *name;

	/** Test-case function. */
	pcut_test_func_t test_func;

	/** Set-up function of a suite. */
	pcut_setup_func_t setup_func;
	/** Tear-down function of a suite. */
	pcut_setup_func_t teardown_func;

	/** Extra attributes. */
	pcut_extra_t *extras;

	/** Extra attributes for main() function. */
	pcut_main_extra_t *main_extras;

	/** Nested lists. */
	pcut_item_t *nested;
};

/** @endcond */

#endif
