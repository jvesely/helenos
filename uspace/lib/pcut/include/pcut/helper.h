/*
 * Copyright (c) 2014 Vojtech Horky
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * - Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following disclaimer in the
 *   documentation and/or other materials provided with the distribution.
 * - The name of the author may not be used to endorse or promote products
 *   derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/** @file
 * Helper macros.
 */
#ifndef PCUT_HELPER_H_GUARD
#define PCUT_HELPER_H_GUARD

/** @cond devel */

/** Join the two arguments on preprocessor level (inner call). */
#define PCUT_JOIN_IMPL(a, b) a##b

/** Join the two arguments on preprocessor level. */
#define PCUT_JOIN(a, b) PCUT_JOIN_IMPL(a, b)

/** Quote the parameter (inner call). */
#define PCUT_QUOTE_IMPL(x) #x

/** Quote the parameter. */
#define PCUT_QUOTE(x) PCUT_QUOTE_IMPL(x)

/** Get first argument from macro var-args. */
#define PCUT_VARG_GET_FIRST(x, ...) x

/** Get all but first arguments from macro var-args. */
#define PCUT_VARG_SKIP_FIRST(x, ...) __VA_ARGS__

#endif
