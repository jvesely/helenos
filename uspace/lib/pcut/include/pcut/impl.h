/*
 * Copyright (c) 2012-2013 Vojtech Horky
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * - Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following disclaimer in the
 *   documentation and/or other materials provided with the distribution.
 * - The name of the author may not be used to endorse or promote products
 *   derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#if !defined(PCUT_TEST_H_GUARD) && !defined(PCUT_INTERNAL)
#error "You cannot include this file directly."
#endif

#ifndef PCUT_IMPL_H_GUARD
#define PCUT_IMPL_H_GUARD

#include "prevs.h"
#include <stdlib.h>

enum {
	PCUT_KIND_SKIP,
	PCUT_KIND_NESTED,
	PCUT_KIND_SETUP,
	PCUT_KIND_TEARDOWN,
	PCUT_KIND_TESTSUITE,
	PCUT_KIND_TEST
};

typedef struct pcut_item pcut_item_t;
typedef void (*pcut_test_func_t)(void);
typedef void (*pcut_setup_func_t)(void);

struct pcut_item {
	pcut_item_t *previous;
	pcut_item_t *next;
	int id;
	int kind;
	union {
		struct {
			const char *name;
			pcut_setup_func_t setup;
			pcut_setup_func_t teardown;
		} suite;
		struct {
			const char *name;
			pcut_test_func_t func;
		} test;
		/* setup is used for both set-up and tear-down */
		struct {
			pcut_setup_func_t func;
		} setup;
		struct {
			pcut_item_t *last;
		} nested;
		struct {
			int dummy;
		} meta;
	};
};

void pcut_failed_assertion(const char *message);
void pcut_failed_assertion_fmt(const char *fmt, ...);
int pcut_str_equals(const char *a, const char *b);
int pcut_main(pcut_item_t *last, int argc, char *argv[]);

#define PCUT_ASSERTION_FAILED(fmt, ...) \
	pcut_failed_assertion_fmt(__FILE__ ":%d: " fmt, __LINE__, ##__VA_ARGS__)

#define PCUT_JOIN_IMPL(a, b) a##b
#define PCUT_JOIN(a, b) PCUT_JOIN_IMPL(a, b)

#define PCUT_ITEM_NAME(number) \
	PCUT_JOIN(pcut_item_, number)

#define PCUT_ITEM_NAME_PREV(number) \
	PCUT_JOIN(pcut_item_, PCUT_JOIN(PCUT_PREV_, number))

#define PCUT_ADD_ITEM(number, itemkind, ...) \
		static pcut_item_t PCUT_ITEM_NAME(number) = { \
				.previous = &PCUT_ITEM_NAME_PREV(number), \
				.next = NULL, \
				.id = -1, \
				.kind = itemkind, \
				__VA_ARGS__ \
		}; \

#define PCUT_TEST_IMPL(testname, number) \
		static void PCUT_JOIN(test_, testname)(void); \
		PCUT_ADD_ITEM(number, PCUT_KIND_TEST, \
				.test = { \
					.name = #testname, \
					.func = PCUT_JOIN(test_, testname) \
				} \
		) \
		void PCUT_JOIN(test_, testname)(void)

#define PCUT_TEST_SUITE_IMPL(suitename, number) \
		PCUT_ADD_ITEM(number, PCUT_KIND_TESTSUITE, \
				.suite = { \
					.name = #suitename, \
					.setup = NULL, \
					.teardown = NULL \
				} \
		)

#define PCUT_TEST_BEFORE_IMPL(number) \
		static void PCUT_JOIN(setup_, number)(void); \
		PCUT_ADD_ITEM(number, PCUT_KIND_SETUP, \
				.setup.func = PCUT_JOIN(setup_, number) \
		) \
		void PCUT_JOIN(setup_, number)(void)

#define PCUT_TEST_AFTER_IMPL(number) \
		static void PCUT_JOIN(teardown_, number)(void); \
		PCUT_ADD_ITEM(number, PCUT_KIND_TEARDOWN, \
				.setup.func = PCUT_JOIN(teardown_, number) \
		) \
		void PCUT_JOIN(teardown_, number)(void)

#define PCUT_EXPORT_IMPL(identifier, number) \
	pcut_item_t pcut_exported_##identifier = { \
		.previous = &PCUT_ITEM_NAME_PREV(number), \
		.next = NULL, \
		.kind = PCUT_KIND_SKIP \
	}

/*
 * Trick with [] and & copied from
 * http://bytes.com/topic/c/answers/553555-initializer-element-not-constant#post2159846
 * because following does not work:
 * extern int *a;
 * int *b = a;
 */
#define PCUT_IMPORT_IMPL(identifier, number) \
	extern pcut_item_t pcut_exported_##identifier; \
	PCUT_ADD_ITEM(number, PCUT_KIND_NESTED, \
		.nested.last = &pcut_exported_##identifier \
	)

#define PCUT_INIT_IMPL(first_number) \
	static pcut_item_t PCUT_ITEM_NAME(__COUNTER__) = { \
		.previous = NULL, \
		.next = NULL, \
		.id = -1, \
		.kind = PCUT_KIND_SKIP \
	}; \
	PCUT_TEST_SUITE(Default);

#define PCUT_MAIN_IMPL(last_number) \
	static pcut_item_t pcut_item_last = { \
		.previous = &PCUT_JOIN(pcut_item_, PCUT_JOIN(PCUT_PREV_, last_number)), \
		.kind = PCUT_KIND_SKIP \
	}; \
	int main(int argc, char *argv[]) { \
		return pcut_main(&pcut_item_last, argc, argv); \
	}

#ifdef PCUT_DEBUG_BUILD
#define PCUT_DEBUG(msg, ...) \
	printf("[PCUT]: Debug: " msg "\n", ##__VA_ARGS__)
#else
#define PCUT_DEBUG(msg, ...) (void)0
#endif

#endif
