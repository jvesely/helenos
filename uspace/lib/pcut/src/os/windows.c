/*
 * Copyright (c) 2014 Vojtech Horky
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * - Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following disclaimer in the
 *   documentation and/or other materials provided with the distribution.
 * - The name of the author may not be used to endorse or promote products
 *   derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/** @file
 *
 * Windows-specific functions for test execution via the popen() system call.
 */

/*
 * Code inspired by Creating a Child Process with Redirected Input and Output:
 * http://msdn.microsoft.com/en-us/library/ms682499%28VS.85%29.aspx
 */

#include "../internal.h"

#include <windows.h>
#include <tchar.h>
#include <stdio.h>
#include <strsafe.h>


/** Maximum size of stdout we are able to capture. */
#define OUTPUT_BUFFER_SIZE 8192

/** Maximum command-line length. */
#define PCUT_COMMAND_LINE_BUFFER_SIZE 256

/** Buffer for assertion and other error messages. */
static char error_message_buffer[OUTPUT_BUFFER_SIZE];

/** Buffer for stdout from the test. */
static char extra_output_buffer[OUTPUT_BUFFER_SIZE];

/** Prepare for a new test.
 *
 * @param test Test that is about to be run.
 */
static void before_test_start(pcut_item_t *test) {
	pcut_report_test_start(test);

	memset(error_message_buffer, 0, OUTPUT_BUFFER_SIZE);
	memset(extra_output_buffer, 0, OUTPUT_BUFFER_SIZE);
}

/** Report that a certain function failed.
 *
 * @param test Current test.
 * @param failed_function_name Name of the failed function.
 */
static void report_func_fail(pcut_item_t *test, const char *failed_function_name) {
	/* TODO: get error description. */
	sprintf_s(error_message_buffer, OUTPUT_BUFFER_SIZE - 1,
		"%s failed: %s.", failed_function_name, "unknown reason");
	pcut_report_test_done(test, TEST_OUTCOME_ERROR, error_message_buffer, NULL, NULL);
}

/** Read full buffer from given file descriptor.
 *
 * This function exists to overcome the possibility that read() may
 * not fill the full length of the provided buffer even when EOF is
 * not reached.
 *
 * @param fd Opened file descriptor.
 * @param buffer Buffer to store data into.
 * @param buffer_size Size of the @p buffer in bytes.
 * @return Number of actually read bytes.
 */
static size_t read_all(HANDLE fd, char *buffer, size_t buffer_size) {
	DWORD actually_read;
	char *buffer_start = buffer;
	BOOL okay = FALSE;

	do {
		okay = ReadFile(fd, buffer, buffer_size, &actually_read, NULL);
		if (!okay) {
			break;
		}
		if (actually_read > 0) {
			buffer += actually_read;
			buffer_size -= actually_read;
			if (buffer_size == 0) {
				break;
			}
		}
	} while (actually_read > 0);
	if (buffer_start != buffer) {
		if (*(buffer - 1) == 10) {
			*(buffer - 1) = 0;
			buffer--;
		}
	}
	return buffer - buffer_start;
}

struct test_output_data {
	HANDLE pipe_stdout;
	HANDLE pipe_stderr;
	char *output_buffer;
	size_t output_buffer_size;
};

static DWORD WINAPI read_test_output_on_background(LPVOID test_output_data_ptr) {
	size_t stderr_size = 0;
	struct test_output_data *test_output_data = (struct test_output_data *) test_output_data_ptr;

	stderr_size = read_all(test_output_data->pipe_stderr,
		test_output_data->output_buffer,
		test_output_data->output_buffer_size - 1);
	read_all(test_output_data->pipe_stdout,
		test_output_data->output_buffer,
		test_output_data->output_buffer_size - 1 - stderr_size);

	return 0;
}

/** Run the test as a new process and report the result.
 *
 * @param self_path Path to itself, that is to current binary.
 * @param test Test to be run.
 */
void pcut_run_test_forking(const char *self_path, pcut_item_t *test) {
	/* TODO: clean-up if something goes wrong "in the middle" */
	BOOL okay = FALSE;
	DWORD rc;
	int outcome;
	int timed_out;
	int time_out_millis;
	SECURITY_ATTRIBUTES security_attributes;
	HANDLE link_stdout[2] = { NULL, NULL };
	HANDLE link_stderr[2] = { NULL, NULL };
	HANDLE link_stdin[2] = { NULL, NULL };
	PROCESS_INFORMATION process_info;
	STARTUPINFO start_info;
	char command[PCUT_COMMAND_LINE_BUFFER_SIZE];
	struct test_output_data test_output_data;
	HANDLE test_output_thread_reader;


	before_test_start(test);

	/* Pipe handles are inherited. */
	security_attributes.nLength = sizeof(SECURITY_ATTRIBUTES);
	security_attributes.bInheritHandle = TRUE;
	security_attributes.lpSecurityDescriptor = NULL;

	/* Create pipe for stdout, make sure it is not inherited. */
	okay = CreatePipe(&link_stdout[0], &link_stdout[1], &security_attributes, 0);
	if (!okay) {
		report_func_fail(test, "CreatePipe(/* stdout */)");
		return;
	}
	okay = SetHandleInformation(link_stdout[0], HANDLE_FLAG_INHERIT, 0);
	if (!okay) {
		report_func_fail(test, "SetHandleInformation(/* stdout */)");
		return;
	}

	/* Create pipe for stderr, make sure it is not inherited. */
	okay = CreatePipe(&link_stderr[0], &link_stderr[1], &security_attributes, 0);
	if (!okay) {
		report_func_fail(test, "CreatePipe(/* stderr */)");
		return;
	}
	okay = SetHandleInformation(link_stderr[0], HANDLE_FLAG_INHERIT, 0);
	if (!okay) {
		report_func_fail(test, "SetHandleInformation(/* stderr */)");
		return;
	}

	/* Create pipe for stdin, make sure it is not inherited. */
	okay = CreatePipe(&link_stdin[0], &link_stdin[1], &security_attributes, 0);
	if (!okay) {
		report_func_fail(test, "CreatePipe(/* stdin */)");
		return;
	}
	okay = SetHandleInformation(link_stdin[1], HANDLE_FLAG_INHERIT, 0);
	if (!okay) {
		report_func_fail(test, "SetHandleInformation(/* stdin */)");
		return;
	}

	/* Prepare information for the child process. */
	ZeroMemory(&process_info, sizeof(PROCESS_INFORMATION));
	ZeroMemory(&start_info, sizeof(STARTUPINFO));
	start_info.cb = sizeof(STARTUPINFO);
	start_info.hStdError = link_stderr[1];
	start_info.hStdOutput = link_stdout[1];
	start_info.hStdInput = link_stdin[0];
	start_info.dwFlags |= STARTF_USESTDHANDLES;

	/* Format the command line. */
	sprintf_s(command, PCUT_COMMAND_LINE_BUFFER_SIZE - 1,
		"\"%s\" -t%d", self_path, test->id);

	/* Run the process. */
	okay = CreateProcess(NULL, command, NULL, NULL, TRUE, 0, NULL, NULL,
		&start_info, &process_info);

	if (!okay) {
		report_func_fail(test, "CreateProcess()");
		return;
	}

	// FIXME: kill the process on error

	/* Close handles to the first thread. */
	CloseHandle(process_info.hThread);

	/* Close the other ends of the pipes. */
	okay = CloseHandle(link_stdout[1]);
	if (!okay) {
		report_func_fail(test, "CloseHandle(/* stdout */)");
		return;
	}
	okay = CloseHandle(link_stderr[1]);
	if (!okay) {
		report_func_fail(test, "CloseHandle(/* stderr */)");
		return;
	}
	okay = CloseHandle(link_stdin[0]);
	if (!okay) {
		report_func_fail(test, "CloseHandle(/* stdin */)");
		return;
	}

	/*
	 * Read data from stdout and stderr.
	 * We need to do this in a separate thread to allow the
	 * time-out to work correctly.
	 * Probably, this can be done with asynchronous I/O but
	 * this works for now pretty well.
	 */
	test_output_data.pipe_stderr = link_stderr[0];
	test_output_data.pipe_stdout = link_stdout[0];
	test_output_data.output_buffer = extra_output_buffer;
	test_output_data.output_buffer_size = OUTPUT_BUFFER_SIZE;

	test_output_thread_reader = CreateThread(NULL, 0,
		read_test_output_on_background, &test_output_data,
		0, NULL);

	if (test_output_thread_reader == NULL) {
		report_func_fail(test, "CreateThread(/* read test stdout */)");
		return;
	}

	/* Wait for the process to terminate. */
	timed_out = 0;
	time_out_millis = pcut_get_test_timeout(test) * 1000;
	rc = WaitForSingleObject(process_info.hProcess, time_out_millis);
	PCUT_DEBUG("Waiting for test %s (%dms) returned %d.", test->name, time_out_millis, rc);
	if (rc == WAIT_TIMEOUT) {
		/* We timed-out: kill the process and wait for its termination again. */
		timed_out = 1;
		okay = TerminateProcess(process_info.hProcess, 5);
		if (!okay) {
			report_func_fail(test, "TerminateProcess(/* PROCESS_INFORMATION.hProcess */)");
			return;
		}
		rc = WaitForSingleObject(process_info.hProcess, INFINITE);
	}
	if (rc != WAIT_OBJECT_0) {
		report_func_fail(test, "WaitForSingleObject(/* PROCESS_INFORMATION.hProcess */)");
		return;
	}

	/* Get the return code and convert it to outcome. */
	okay = GetExitCodeProcess(process_info.hProcess, &rc);
	if (!okay) {
		report_func_fail(test, "GetExitCodeProcess()");
		return;
	}

	if (rc == 0) {
		outcome = TEST_OUTCOME_PASS;
	} else if ((rc > 0) && (rc < 10) && !timed_out) {
		outcome = TEST_OUTCOME_FAIL;
	} else {
		outcome = TEST_OUTCOME_ERROR;
	}

	/* Wait for the reader thread (shall be terminated by now). */
	rc = WaitForSingleObject(test_output_thread_reader, INFINITE);
	if (rc != WAIT_OBJECT_0) {
		report_func_fail(test, "WaitForSingleObject(/* stdout reader thread */)");
		return;
	}

	pcut_report_test_done_unparsed(test, outcome, extra_output_buffer, OUTPUT_BUFFER_SIZE);
}

void pcut_hook_before_test(pcut_item_t *test) {
	PCUT_UNUSED(test);

	/*
	 * Prevent displaying the dialog informing the user that the
	 * program unexpectedly failed.
	 */
	SetErrorMode(SEM_FAILCRITICALERRORS | SEM_NOGPFAULTERRORBOX);
}
