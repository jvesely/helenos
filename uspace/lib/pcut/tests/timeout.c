/*
 * Copyright (c) 2012-2013 Vojtech Horky
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * - Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 * - Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following disclaimer in the
 *   documentation and/or other materials provided with the distribution.
 * - The name of the author may not be used to endorse or promote products
 *   derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <pcut/pcut.h>

#ifdef __unix
#include <unistd.h>
#endif
#if defined(__WIN64) || defined(__WIN32) || defined(_WIN32)
#include <windows.h>
#endif

#include <stdio.h>
#include "tested.h"

static void my_sleep(int sec) {
#ifdef __unix
	sleep(sec);
#endif
#if defined(__WIN64) || defined(__WIN32) || defined(_WIN32)
	Sleep(1000 * sec);
#endif
}

PCUT_INIT

PCUT_TEST(shall_time_out) {
	printf("Text before sleeping.\n");
	my_sleep(PCUT_DEFAULT_TEST_TIMEOUT * 5);
	printf("Text after the sleep.\n");
}

PCUT_TEST(custom_time_out,
		PCUT_TEST_SET_TIMEOUT(PCUT_DEFAULT_TEST_TIMEOUT * 3)) {
	printf("Text before sleeping.\n");
	my_sleep(PCUT_DEFAULT_TEST_TIMEOUT * 2);
	printf("Text after the sleep.\n");
}

PCUT_MAIN()
