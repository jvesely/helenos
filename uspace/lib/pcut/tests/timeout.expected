1..2
#> Starting suite Default.
not ok 1 shall_time_out aborted
# stdio: Text before sleeping.
ok 2 custom_time_out
# stdio: Text before sleeping.
# stdio: Text after the sleep.
#> Finished suite Default (failed 1 of 2).
